File: Behavioral/Lab20/Typescript/src/Entities/States/estate_state.ts
import { IBuyable, } from "../../Interfaces/buyable";
import { IRentable } from "../../Interfaces/rentable";
import { ISellable } from "../../Interfaces/sellable";
import { Customer } from "../customer";
import { RealEstate } from "../real_estate";

export abstract class EstateState implements IBuyable, ISellable, IRentable
{
    protected _estate: RealEstate;
    constructor(estate: RealEstate)
    {
        this._estate = estate;
    }
    public abstract rent(customer: Customer, startDate: Date, endDate: Date): void;

    public abstract sell(customer: Customer): void;

    public abstract setEstateToBuy(customer: Customer, price: number): void;

}

File: Behavioral/Lab20/C#/Entities/States/EstateState.cs
using System;
using System.Collections.Generic;
using System.Text;
using C_.Entities;
using C_.Interfaces;

namespace C_.Entities.States
{
    public abstract class EstateState : IBuyable, ISellable, IRentable
    {
        protected RealEstate _estate;
        public EstateState(RealEstate estate)
        {
            _estate = estate;
        }
        public abstract void Rent(Customer customer, DateTime startDate, DateTime endDate);

        public abstract void Sell(Customer customer);

        public abstract void SetEstateToBuy(Customer customer, double price);

    }
}
